source/common/grpc/grpc_web_filter.cc

```diff
@@ -0,0 +1,109 @@
+  }
+  headers.insertContentType().value(Http::Headers::get().ContentTypeValues.Grpc);
+
+  const Http::HeaderEntry* accept = headers.get(Http::Headers::get().Accept);
+  if (accept != nullptr &&
+      Http::Headers::get().ContentTypeValues.GrpcWebText == accept->value().c_str()) {
+    is_text_response_ = true;
+  }
+  headers.addStatic(Http::Headers::get().TE, Http::Headers::get().TEValues.Trailers);
+  headers.addStatic(Http::Headers::get().GrpcAcceptEncoding,
+                    Http::Headers::get().GrpcAcceptEncodingValues.Default);
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus GrpcWebFilter::decodeData(Buffer::Instance& data, bool) {
+  if (!is_text_request_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // Parse application/grpc-web-text format.
+  if (data.length() + decoding_buffer_.length() < 4) {
+    decoding_buffer_.move(data);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t needed =
+      (data.length() + decoding_buffer_.length()) / 4 * 4 - decoding_buffer_.length();
+  decoding_buffer_.move(data, needed);
+  const std::string decoded = Base64::decode(
+      std::string(static_cast<const char*>(decoding_buffer_.linearize(decoding_buffer_.length())),
+                  decoding_buffer_.length()));
+  decoding_buffer_.drain(decoding_buffer_.length());
+  decoding_buffer_.move(data);
+  data.add(decoded);
+  return Http::FilterDataStatus::Continue;
+}
+
+// Implements StreamEncoderFilter.
+Http::FilterHeadersStatus GrpcWebFilter::encodeHeaders(Http::HeaderMap& headers, bool) {
+  if (is_text_response_) {
+    headers.insertContentType().value(Http::Headers::get().ContentTypeValues.GrpcWebText);
+  } else {
+    headers.insertContentType().value(Http::Headers::get().ContentTypeValues.GrpcWeb);
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool) {
+  if (!is_text_response_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // Encodes the response as base64.
+  std::vector<Frame> frames;
+  decoder_.decode(data, frames);
+  for (auto& frame : frames) {
+    Buffer::OwnedImpl temp;
+    temp.add(&frame.flags_, 1);
+    const uint32_t length = htonl(frame.length_);
+    temp.add(&length, 4);
+    if (frame.length_ > 0) {
+      temp.add(*frame.data_);
+    }
+    data.add(Base64::encode(temp, temp.length()));
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus GrpcWebFilter::encodeTrailers(Http::HeaderMap& trailers) {
+  Buffer::OwnedImpl temp;
+  trailers.iterate([](const Http::HeaderEntry& header, void* context) -> void {
+    Buffer::Instance* temp = static_cast<Buffer::Instance*>(context);
+    temp->add(header.key().c_str(), header.key().size());
+    temp->add(":");
+    temp->add(header.value().c_str(), header.value().size());
+    temp->add("\r\n");
+  }, &temp);
+  Buffer::OwnedImpl buffer;
+  buffer.add(&GRPC_WEB_TRAILER, 1);
```

Comment:
It's not in gRPC spec. Instead, it's defined in gRPC-Web. Trailers are encoded in a frame denoted with 0b1000000 + 4 bytes length in network order. The frame data is CRLF separated headers.